Convert Array4 to ArrayND

ankithadas opened the pull request:
## Summary

This PR introduces a preliminary implementation that generalizes Array4 to an N-dimensional case, extending AMReX support for arbitrary dimensionality as outlined in #3955.

The new `ArrayND` implementation has the same size as `Array4` (64 bytes) and is intended to exhibit identical behaviour. However, this change modifies the types of `begin` and `end` and removes `ncomp`, which may break downstream code. In particular, `pyamrex` will likely need to be updated to accommodate these type changes. All internal AMReX source code that uses `begin`, `end`, or `ncomp` has been updated.

One open question concerns the behaviour of `ArrayND<dim>::operator()(idx...)` when the number of indices provided is `dim-1`. In `Array4`, this case sets the last dimension `ncomp` to zero. Should this behaviour be extended to all `dim` cases in `ArrayND` ? 

## Additional background

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [x] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

ankithadas committed:
Generalization of Array4 to ArrayND

ankithadas committed:
Fix Clang-Tidy errors. Refactor SFINAE

ankithadas committed:
Fix constructor constexpr. std::array should be compatible with Gpu. Tests should show.

WeiqunZhang commented:
/run-hpsf-gitlab-ci

zingale commented:
does `std::array` work with HIP on device?

BenWibking commented:
> does `std::array` work with HIP on device?

We use it. I haven't seen any issues.

AlexanderSinn commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +57,373 @@ namespace amrex {
+            : p(a_p), begin(a_begin.x, a_begin.y, a_begin.z, 0), end(a_end.x, a_end.y, a_end.z, a_ncomp)
+        {
+            Long current_stride = 1;
+            constexpr_for<0, dim-1>([&] (int d) {
```
I think it would be better to use regular for loops. Since the bounds are known at compile time, all compilers should fully unroll it anyway. constexpr_for could be a bit heavy to compile.

AlexanderSinn commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +57,373 @@ namespace amrex {
 #if defined(AMREX_DEBUG) || defined(AMREX_BOUND_CHECK)
-            index_assert(i,j,k,0);
+            if constexpr (sizeof...(i) == dim) {
+                index_assert(i...);
```
You can put the `index_assert` inside `get_offset` to avoid needing so many ifdefs.

ankithadas replied:
I have currently kept it seperate to make `get_offset` method constexpr.

AlexanderSinn commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +57,373 @@ namespace amrex {
+            if constexpr (sizeof...(i) == dim) {
+                index_assert(i...);
+            } else if constexpr (sizeof...(i) == dim - 1) {
+                // When number of indices is one less than dimension, we assume the last index is zero
```
In this case the last index should not be zero but effectively `begin[dim-1]`. If it were zero, the index calculation would still add instructions to subtract `begin[dim-1]`, multiply by the stride, and add the result to the offset. Currently with Array4 this is not done. To guarantee the absence of the extra instructions, there should be two versions of `get_offset` where one of them leaves out the calculation for the last index.

AlexanderSinn commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +57,373 @@ namespace amrex {
+#elif (AMREX_SPACEDIM == 3)
+            return this->operator()(iv[0],iv[1],iv[2],0);
 #endif
         }
```
Don't these two definitions collide? Ideally, this would be one function without needing the special case.

AlexanderSinn commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +57,373 @@ namespace amrex {
+
+        template <typename... idx, std::enable_if_t<sizeof...(idx) == dim,int> = 0>
+        [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        constexpr Long get_offset (idx... indices) const noexcept
```
I think we can simplify this and remove this version of `get_offset`by constructing an IntVectND at the place where it is called. Then `get_impl` can also be removed.

AlexanderSinn reviewed (commented):
Currently there are a lot of special cases for `dim == 4` for both the accessors and constructors. I think in general it is useful to use an ArrayND with n-1 spatial dimensions and 1 component dimension, even if `dim != 4`. For example, I would use this for a 2+1D array. There should be these 5 accessor overloads, each for both `operator()` and `ptr()`:
- `idx... i`: This should cover both `sizeof...(idx) == dim` and `dim-1` like currently, but also replace the `int, int, int, int` version. For this, we might have to allow all types that are implicitly convertible to int for `idx`.
- `IntVectND<Dim>`:  for `Dim == dim`, `dim-1` and `DIM < dim-1`. In the last case, the remaining indexes are set to 0, except the last one, which is set to `lo[dim-1]` but not present in the offset calculation.
- `IntVectND<Dim>, int`: for `Dim == dim-1` and `DIM < dim-1`. In the second case, the IntVect is zero-extended as above. This overload won't be usable if `dim == 1`
- `Dim3` kept for backward compatibility, only works if `dim == 4`.
- `Dim3, int` kept for backward compatibility, only works if `dim == 4`.

In addition to that, I think there should be more constructors, for example, taking a BoxND as input or an ArrayND of different dimensions.

AlexanderSinn commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +57,373 @@ namespace amrex {
-            return this->operator()(iv[0],iv[1],iv[2]);
+            return this->operator()(iv[0],iv[1],0,0);
+#elif (AMREX_SPACEDIM == 3)
+            return this->operator()(iv[0],iv[1],iv[2],0);
```
@WeiqunZhang  Would it be ok to put `begin` here instead of zero? That would be more consistent with the component index and would allow for a small performance optimization, as the offset resulting from that index would be zero.

ankithadas committed:
Updated operator() and ptr() methods. Fixed error message for index out of bounds for GPU.

ankithadas committed:
Added ptr() overload for Dim3 (dim==4) case.

ankithadas committed:
Removed constexpr_for in index_assert

ankithadas commented:
@AlexanderSinn  Could you provide more information regarding the types of constructors that would be beneficial for your application?

I am not convinced that `BoxND` as an input is a good idea. There is an implicit understanding that `BoxND` is related to `AMREX_SPACEDIM` of the application, whereas `ArrayND` is not necessarily tied to this. This mismatch can lead to unexpected behaviour.

ankithadas committed:
Minor Refactor. Updated ubound, lbound, and operator<< for ArrayND.

AlexanderSinn commented:
Yes, currently lower dimensions also use Array4. I assume this is out of simplicity and to make it easy to write code that works in 1D, 2D and 3D by always passing i,j,k,n to the array. However, I think eventually it should also be possible to use the ArrayND of the dimension the code is compiled for (plus one), as this should be slightly more performant and cleaner. The constructors should allow for both ways of using the array and should essentially mirror the interface of `operator()` for what combination of inputs and dimensions are allowed. For constructors using BoxND I would suggest:

- `T*, BoxND<Dim>`:  for `Dim == dim`, `dim-1` and `DIM < dim-1`. In the last case, the remaining dimensions are set to begin=0, end=1.
- `T*, BoxND<Dim>, int ncomp`: for `Dim == dim-1` and `DIM < dim-1`. In the second case, the Box is extended with begin=0, end=1. This overload won't be usable if `dim == 1`

Additionally, the `ArrayND, ncomp` and `ArrayND, start_comp, ncomp` constructors could be made to work with `N>=2` instead of only `N==4`.

For the application I'm working on, I currently use implicit conversions from Array4 to a 2D array and a 3D array, where the 3D version removes the second to last dimension but keeps the last one. Dimensions that are removed are asserted to have size one if the constructor is called on the CPU.  This might be best left to the application, not AMReX.

ankithadas committed:
Added additional constructors for Array4

ankithadas commented:
I believe the final issue to address is the default indices used by `ArrayND::operator()(IntVect<Dim>)` when `Dim < N`. In my opinion, the default value should be `begin` rather than zeros.

ankithadas committed:
Fix sycl tests. Removed Dim variable name as it does not indicate dimensions.

AlexanderSinn commented:
The forward declaration
```C++
template <typename T> struct Array4;
```
in `AMReX_BaseFwd.H` still needs to be updated.

Also in the `ArrayND, ncomp` constructors, replace `3` with `N-1`.

WeiqunZhang commented on Src/Particle/AMReX_ParticleContainerI.H:
```diff
@@ -976,9 +976,9 @@ ParticleContainer_impl<ParticleType, NArrayReal, NArrayInt, Allocator, CellAssig
-                    +(j-imf_arr.begin.y)*imf_arr.jstride
-                    +(k-imf_arr.begin.z)*imf_arr.kstride)
+                    ((i-imf_arr.begin[0])
+                    +(j-imf_arr.begin[1])*imf_arr.nstride[1]
```
Should this be `nstride[0]`? A similar issue in the line below.

ankithadas replied:
Oh yeah good catch.

WeiqunZhang commented:
@yut23 Note that this PR will break the GDB script (https://github.com/AMReX-Codes/amrex/tree/development/Tools/GDB) contributed by you. It would be great if you can fix it when this is merged.

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +61,453 @@ namespace amrex {
-    struct Array4
+    namespace detail {
+        /**
+         * /brief Indices checker for ArrayND
```
If this is intended for doxygen, it should be `\brief`, not `/brief`.

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +61,453 @@ namespace amrex {
-        {}
+              end(rhs.end)
+        {
+            begin[3] = 0;
```
@AlexanderSinn has already mentioned this. `3` should be replaced by `N-1`.

zingale commented:
one of our Castro problem setups fails to build with this PR:
http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2026-01-05-003/rad-2Tshock-3d.make.out

WeiqunZhang commented:
The Castro issue is that the indices passed to ArrayND are Real not integers.
```
    Real jfac = 1.0_rt;
    Real kfac = 1.0_rt;
    fine(i,j,k) = crse(clo.x, j / jfac, k / kfac) / rfac;
```

ankithadas commented:
> one of our Castro problem setups fails to build with this PR: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2026-01-05-003/rad-2Tshock-3d.make.out

The error is from this line 
```(const int&, amrex::Real, amrex::Real)â€™
  327 |                 fine(i,j,k) = crse(clo.x, j / jfac, k / kfac) / rfac;
```

I believe `jfac` is of type `amex::Real`. Previously this code relied on implicit conversion from `Real` to `int` when indexing `Array4`. The changes in this PR now require all index arguments to be explicitly `int`.

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +61,453 @@ namespace amrex {
+        struct ArrayNDIndexCheck_impl {
+            constexpr static bool value =
+                (sizeof...(idx) == N || sizeof...(idx) == N - 1)
+                && (Conjunction<std::is_same<int, std::decay_t<idx>>...>::value);
```
We might want to relax this to idx is an integral type rather than it must be int.

AlexanderSinn replied:
`IsConvertible_v<int, idx...>`  from `AMReX_TypeTraits.H` could be used for this. It will check if an implicit conversion is allowed when calling a function with that type.

WeiqunZhang replied:
I think we should forbid implicit conversion from Real.

ankithadas replied:
My reasoning was to prevent implicit type casting. This would prevent users from using `long` or `long long` as indexing arguments. I can change the constraint such that the index arguments have size smaller than or equal to int

WeiqunZhang replied:
Restrict it to signed integer with size <= sizeof(int) is fine. But we also want to support enum types. We have code doing this. I think it's a style that we should support.
```
namespace EddyDiff {
    enum {
        Mom_h = 0,
        Theta_h,
        KE_h,
        Scalar_h,
        Q_h,
        Mom_v,
        Theta_v,
        KE_v,
        Scalar_v,
        Q_v,
        Turb_lengthscale,
        NumDiffs
    };
}

a(i,j,k,EddyDiff::KE_v)
```

WeiqunZhang replied:
We could try to make sure enum is only allowed for the last index.

ankithadas commented:
> I believe the final issue to address is the default indices used by `ArrayND::operator()(IntVect<Dim>)` when `Dim < N`. In my opinion, the default value should be `begin` rather than zeros.

@AlexanderSinn agrees that missing indices should default to `begin`. Any thoughts @WeiqunZhang ?

zingale commented:
I'm fine with modifying that part of Castro to explicitly cast to int.

WeiqunZhang commented:
> > I believe the final issue to address is the default indices used by `ArrayND::operator()(IntVect<Dim>)` when `Dim < N`. In my opinion, the default value should be `begin` rather than zeros.
> 
> @AlexanderSinn agrees that missing indices should default to `begin`. Any thoughts @WeiqunZhang ?

Yes, I also agree.

ankithadas committed:
Bug fixes and updates to indexing operation. Changed index type checking to support integral types that can be safely converted to integer and allow enum as last index value.

ankithadas committed:
Updated BaseFwd

ankithadas committed:
Added Doxygen documentation

ankithadas commented:
I am not quite sure why documentation build failed. Any insights ?

WeiqunZhang commented:
```
2026-01-07T00:31:56.6866927Z Generating XML output for class amrex_paralleldescriptor_/home/runner/work/amrex/amrex/Src/Base/AMReX_Array4.H:238: warning: Found unknown command '\TODO'
2026-01-07T00:31:56.6867503Z /home/runner/work/amrex/amrex/Src/Base/AMReX_Array4.H:399: warning: argument 'i...' of command @param is not found in the argument list of amrex::ArrayND< T, N >::operator()(idx... i) const
2026-01-07T00:31:56.6867798Z /home/runner/work/amrex/amrex/Src/Base/AMReX_Array4.H:405: warning: Found unknown command '\TODO'
2026-01-07T00:31:56.6868451Z /home/runner/work/amrex/amrex/Src/Base/AMReX_Array4.H:399: warning: The following parameter of amrex::ArrayND::operator()(idx... i) const is not documented:
2026-01-07T00:31:56.6868580Z   parameter 'i'
2026-01-07T00:31:56.6869111Z /home/runner/work/amrex/amrex/Src/Base/AMReX_Array4.H:476: warning: argument 'i...' of command @param is not found in the argument list of amrex::ArrayND< T, N >::ptr(idx... i) const
2026-01-07T00:31:56.6869602Z /home/runner/work/amrex/amrex/Src/Base/AMReX_Array4.H:476: warning: The following parameter of amrex::ArrayND::ptr(idx... i) const is not documented:
2026-01-07T00:31:56.6869704Z   parameter 'i'
2026-01-07T00:31:56.6869812Z module::amrex_pd_bcast
```

ankithadas committed:
Documentation fix

ankithadas committed:
Spell check

AlexanderSinn commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +61,613 @@ namespace amrex {
+         * This constructor specifies begin and end indices using IntVectND<N> for all N dimensions.
+         */
+        AMREX_GPU_HOST_DEVICE
+        constexpr ArrayND (T* a_p, IntVectND<N> const& a_smallend, IntVectND<N> const& a_bigend) noexcept
```
This has the wrong name. Usually in amrex, bigend refers to an inclusive high index. Just begin and end would be better names for this. 

AlexanderSinn replied:
Or maybe it is better if this actually were an inclusive bigend

ankithadas committed:
Update variable names

ankithadas commented:
@yut23 I have completed a preliminary implementation of updated GDB scripts for `ArrayND` (https://github.com/ankithadas/amrex/tree/ArrayND-GDB-Tools). This seems to work for `ArrayND::operator()(i...)`, but I'm unsure how to access template parameters for member functions. Feel free to use this as a starting point for your PR

AlexanderSinn commented:
Hmm, `amrex::ArrayND<T, 3>` looks to make GPU kernels 3-5% slower compared to the 3D array that I previously used. Also, one place using Array4 directly is slower with this PR. Looking at some examples in Compiler Explorer suggests that subtracting begin from the ncomp variable prevents some simplifications when having multiple array accesses.

ankithadas commented:
Could you share details of your 3D array implementation? A simple way to make the computational cost the same is by adding specific overload for `ArrayND<T,4>` that removes the `begin[3] - end[3]` for `ncomp`. A general approach would be adding another class template parameter asking the user if the last dimension is `ncomp`. In that case, we could simplify the offset calculations.

AlexanderSinn commented:
It is the Array2 and (mainly) Array3 that I replaced with ArrayND here https://github.com/Hi-PACE/hipace/pull/1336/files#diff-32376acddbd6018a475c8f547ae49c6087080c5470cc4cf42718060b2295c7b9
After doing more benchmarking, it now seems that most of the performance difference comes from the general way I calculate the offset and not just from begin[3]. I use a 64-bit offset variable that combines all the begins. However, I think this approach cannot be used in amrex as it would increase the size of Array4 (unless begin and end are removed in non-debug builds like in hipace, but these variables are sometimes used by other code) and because it could overflow the 64 bit index when using it with many levels of mesh refinement at the edge of a big 3D domain.

AlexanderSinn commented:
As for adding another template parameter, I would do it by reinterpreting Array4 as a 3, 1 Array. The arrays in hipace for example, would be 2, 1 and 2, 0 in that notation. Not sure if this ends up being cleaner or more complicated in the end.

WeiqunZhang commented:
Let's keep it simple. Having two size template parameters seems confusing.

ankithadas committed:
Bug fix. Didn't change offset calculation when M < N.

ankithadas committed:
Added last_dim_componet template paramter

ankithadas committed:
Update BaseFwd

WeiqunZhang commented:
I have run some tests on Perlmutter and Frontier, and have not noticed any performance difference between this branch and the development. So that's good.

AlexanderSinn commented:
Did you test ArrayND with or without the "Added last_dim_componet template paramter" commit?

WeiqunZhang commented:
I was talking about the latest version with the new bool template parameter. I also tried the version before the change. Most of the tests did not show much difference, but one test was 5-10% slower with the old version.

ankithadas commented:
Any idea why the Intel SYCL tests fail when using standard `for` loops instead of `constexpr_for`?

WeiqunZhang commented:
I think it was just some large kernels were hitting the hardware and/or software limit of Intel GPUs.

WeiqunZhang commented:
Looks like we will need to add `-Wno-maybe-uninitialized` to that failed gcc test.

ankithadas committed:
Fix GNU@12 C++17 w/o Fortran [tests]

ankithadas commented:
Not sure why that test fails for some commits and not others. Anyway, I have updated the test

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +61,634 @@ namespace amrex {
-              end(rhs.end),
-              ncomp(rhs.ncomp)
-            {}
+        constexpr ArrayND (ArrayND<std::remove_const_t<T>, N> const& rhs) noexcept
```
I think it will be safer if we make the type of rhs be `ArrayND<std::remove_const_t<T>, N, last_dim_component> const&`.

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +61,634 @@ namespace amrex {
         AMREX_GPU_HOST_DEVICE
-        explicit operator bool() const noexcept { return p != nullptr; }
+        explicit operator bool() const noexcept {
+            return p != nullptr && end.allGT(begin);
```
I understand it's always a tradeoff. But is the second part of the test necessary in most cases if we trust the users? Currently this function is often used like this
```
if (a && a(i,j,k) > 0) { ... }
```
It offers a cheap test. Maybe we could make `bool()` simple and add a new function called `ok` that checks the pointer and the bounds.

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +61,634 @@ namespace amrex {
+        template <int M, class U=T, std::enable_if_t<!std::is_void_v<U>,int> = 0>
         [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-        U& operator() (int i, int j, int k, int n) const noexcept {
+        U& operator() (IntVectND<M> const& iv) const noexcept {
```
This comment applies to all versions of `opearator()` and the constructors. Since we have the `last_dim_component` template parameter now, should we be more strict on the value of `M`? For example, should we only allow either `M == N` or `M == N-1 && last_dim_component`.

WeiqunZhang replied:
My question still remains. Is it worth to allow code like this?
```
ArrayND<Real,4> a(p, IntVectND<2>(0), IntVectND<2>(4));
a(IntVectND<1>(0)) = 4;
```

ankithadas replied:
Looks like I missed this case. Sorry about that. I don’t specifically need this feature in my code, but I can see it being useful for applications with a mix of periodic and non-periodic directions, where this overload could be used to set those components.

That said, if used improperly, this feature could also lead to subtle or unexpected bugs.

WeiqunZhang replied:
Agreed. My feeling is we should make it hard for the users to write incorrect code, even if it means they need to explicitly expand or shrink IntVectND in certain use cases. (We do have helper functions for that.)

ankithadas replied:
Sounds good.

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +61,634 @@ namespace amrex {
+         * For N!=4 case, the user is expected to keep track if they are using the last dimension
+         * as components.
+         */
+        template <int M=N, std::enable_if_t<M==4 || last_dim_component,int> = 0>
```
Do we really need this function if last_dim_component is false?

WeiqunZhang commented:
For discussion. Should we make last_dim_component = false by default and make it true for Array4?

WeiqunZhang commented on Src/Base/AMReX_TypeTraits.H:
```diff
@@ -275,6 +269,30 @@ namespace amrex
+    //! \brief Test if type T is a valid index type. Valid index type is integral and can be safely
+    //!         converted to int without narrowing.
+    template <typename T>
+    struct IsValidIndexType {
```
Since this is not in Array4 or namespace detail, is it a good idea to have `int` hardwired?

ankithadas commented:
> For discussion. Should we make last_dim_component = false by default and make it true for Array4?

I agree. Would an `enum` be better ? Something like this 

```c++
enum class DimSpec: int { Spatial = 0, Component = 1 };
```

WeiqunZhang commented:
`ArrayND<3,DimSpec::Component>` looks weird to me.

ankithadas committed:
Bug fixes for last_dim_component template cases

ankithadas committed:
Fix spellings

ankithadas commented:
I have added explicit overloads for `ArrayND::operator()` and `ArrayND::ptr()` covering the following three argument patterns:
1. `IntVect<N>`  
2. `IntVect<M>` where `1 <= M <= N-1` 
3. `IntVect<M>, int` where `1 <= M <= N-1`  and `last_dim_component=true`

These overloads are not required but were added to provide better error messages. Without them, the compiler errors typically report a failure to find a matching `get_offset` with the given signature, which although correct, does not clearly indicate that the provided arguments fail to meet the required constraint.
I have also added overloads for `Dim3` for `N=3` cases as well. 

Lastly, one issue I noticed is that CTAD cases that worked previously would not work in the new version. The following construction now fails
```c++
Array4 (Real *, IntVect(), IntVect(), int)
```

WeiqunZhang commented:
> Lastly, one issue I noticed is that CTAD cases that worked previously would not work in the new version. The following construction now fails
> 
> ```c++
> Array4 (Real *, IntVect(), IntVect(), int)
> ```
Maybe you were using C++20 previously? I get this with C++17.
```
error: alias template deduction only available with ‘-std=c++20’ or ‘-std=gnu++20’
```
In my test, using C++20 works and `ArrayND` with C++17 works as well.

ankithadas commented:
Sorry, my earlier test case was incorrect, since the `IntVect` overloads were only introduced in this branch. 
The following code fails to compile in the `ArrayND` branch with C++17, but works on the dev branch. Compiling with C++20 fixes the issue. 
```cpp
Real* nullptr_real = nullptr;
Dim3 begin = IntVect(0).dim3();
Dim3 end = IntVect(ncells).dim3();
Array4 t (nullptr_real, begin, end, 1);
```

ankithadas commented:
For discussion. Should we allow this construction now that `last_dim_component` is a template parameter 
```cpp 
ArrayND<Real, 5> a(nullptr, IntVectND<1>(0), IntVectND<1>(ncells));
ArrayND<Real, 5> b(nullptr, BoxND(IntVectND<1>(0), IntVectND<1>(ncells)));
```
I find this rather confusing tbh.

WeiqunZhang commented:
Agreed. It's very confusing. We should only allow `M==N` and `M+1==N && last_dim_component`.

ankithadas committed:
Updated constructors, operator() and ptr()

WeiqunZhang commented:
/run-hpsf-gitlab-ci

WeiqunZhang committed:
Make sure that size() works on uninitialized objects

WeiqunZhang committed:
Make ok() constexpr since it has at least one constexpr path.

WeiqunZhang commented:
This now compiles.
```
constexpr Array4<int> a{};
static_assert(a.size() == 0);
```

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -56,212 +61,630 @@ namespace amrex {
+            });
+            if constexpr (last_dim_component && M == N) {
+                inside = inside && (iv[N-1] >= 0) && (iv[N-1] < end[N-1]);
+            }
```
Could this be simplified to
```suggestion
            constexpr_for<0, M>([&](int d) {
                inside = inside && (iv[d] >= begin[d]) && (iv[d] < end[d]);
            });
```

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -270,46 +693,110 @@ namespace amrex {
+    template <class Tto, class Tfrom, int N>
     [[nodiscard]] AMREX_GPU_HOST_DEVICE
-    Array4<Tto> ToArray4 (Array4<Tfrom> const& a_in) noexcept
+    ArrayND<Tto,N> ToArray4 (ArrayND<Tfrom,N> const& a_in) noexcept
```
I don't think anyone is using this function. We could remove it or we limit it to Array4 only because of its name.

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -270,46 +693,110 @@ namespace amrex {
+        return Dim3{a.begin[0],a.begin[1],a.begin[2]};
+    }
+
+    template <class T, int N, std::enable_if_t<(N != 4), int> = 0>
```
I think we should remove this version of lbound and ubound because of potential confusion.

ankithadas replied:
Would it be better to rename them to ubound_iv ?

WeiqunZhang replied:
I think it's better to not having it. We added the lbound and ubound functions for Array4 because we needed them, not because we anticipated that we would need them. So maybe in the future it will become clear we need a helper function like that for non-Array4 ArrayND.

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -318,19 +805,41 @@ namespace amrex {
-        os << "((" << lbound(a) << ',' << ubound(a) << ")," << a.ncomp << ')';
+    template <class T, int N, std::enable_if_t<(N != 4), int> = 0>
+    [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    IntVectND<N> length (ArrayND<T,N> const& a) noexcept
```
Remove this version of length as well to avoid confusion.

WeiqunZhang commented on Src/Base/AMReX_Array4.H:
```diff
@@ -318,19 +805,41 @@ namespace amrex {
+
+    template <typename T, int N>
+    std::ostream& operator<< (std::ostream& os, const ArrayND<T,N>& a) {
+        os << "((" << lbound(a) << ',' << ubound(a) << "))";
```
```suggestion
    template <typename T, int N, bool C>
    std::ostream& operator<< (std::ostream& os, const ArrayND<T,N,C>& a) {
        os << "(" << a.begin << ',' << a.end-1 << ")";
```

ankithadas committed:
Fix operator<<

Co-authored-by: Weiqun Zhang <WeiqunZhang@lbl.gov>

ankithadas committed:
Implemented suggestions

ankithadas commented:
Thank you for patiently reviewing the PR so many times.

WeiqunZhang commented:
Thank you for the PR! You did the work! If there are still bugs/defects there, it's all my fault for not catching it. :)

zingale commented:
This change causes Castro when built with `DEBUG=TRUE` to run 3x slower.  Is there a way to get the performance back in debug mode?  Things only seem slightly slower when optimized.

WeiqunZhang commented:
Yes we should be able to get the performance back.

On Fri, Jan 16, 2026, 5:31 AM Michael Zingale ***@***.***>
wrote:

> *zingale* left a comment (AMReX-Codes/amrex#4868)
> <https://github.com/AMReX-Codes/amrex/pull/4868#issuecomment-3760039105>
>
> This change causes Castro when built with DEBUG=TRUE to run 3x slower. Is
> there a way to get the performance back in debug mode? Things only seem
> slightly slower when optimized.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AMReX-Codes/amrex/pull/4868#issuecomment-3760039105>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB37TYJUBVO45YJHLDLHFVD4HDR3VAVCNFSM6AAAAACQBZ6IT6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTONRQGAZTSMJQGU>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

WeiqunZhang commented:
I have a version that is only 10% slower than the original version in debug mode.

WeiqunZhang commented:
See #4900
